src/libraries/Microsoft.Extensions.Primitives/src/ChangeToken.cs

```diff
@@ -142,7 +142,7 @@ public void Dispose()
             {
                 // If the previous value is disposable then dispose it, otherwise,
                 // now we've set the disposed sentinel
-                Interlocked.Exchange(ref _disposable, _disposedSentinel).Dispose();
+                Interlocked.Exchange(ref _disposable, _disposedSentinel)!.Dispose();
```

Comment:
Sorry - I was confusing on my previous comment. I think what you had before was correct:

```suggestion
                Interlocked.Exchange(ref _disposable, _disposedSentinel)?.Dispose();
```

It is possible for `_disposable` to be `null` here - when someone creates this class with a `changeTokenProducer` that returns `null`. In that case, nothing will call `SetDisposable` (as far as I can tell), and `_disposable` will still be null when this class is Disposed.

Can you add a unit test for this scenario while we are here? Basically the test is:

```C#
ChangeToken.OnChange(() => null, () => Debug.Fail("Should never be called")).Dispose();
```